### PR TITLE
PC-351: EPC API Integration

### PR DESCRIPTION
**Pull request by chidin194**
What this PR covers:

- All endpoint and page routing in `views.py`, including epc retrieval specifically for Scotland
- Addition of the new `epc-select.html` page
- Changes to `interface.py` to call the new epc endpoints
- A new `epc_api.py` file that handles all the endpoint calls

What needs doing:
 - Sort out use of classes and schemas - **(update 8/12 - this is mainly adding loads/dumps which is probably not a priority so I'm going to leave this for now. I've updated the session schema which I think is more important)**
 - Improve error handling and logging in `epc_api.py`
 - Fix tests
 - Add a package to handle OAuth2
 - Confirm changes to CSV
 - Add UPRN and main heating source to session data
 - Fix bug discussed yesterday

**(Update 8/12) commit pushed today means that the following is left to do:**
 - Use allauth (already in use elsewhere) to handle OAuth2 authentication
 - Fix and add relevant tests, including mocking out the EPC API
 - Replace EPC request URLs and credentials with those for the EPC API prod environment

Also added in today (8/12):
- Making sure the new epc-select page can be viewed in Welsh
- Resetting epc-related session data for manual address look-up as well to avoid the following scenarios:
           - User initially inputs an address for which an EPC is found, goes back and inputs an address for which both calls fail and then has to input address manually
           - User goes from having input an EPC address back to address input and goes directly to the manual address input page
- Logic currently resets all fields we put into the session data from the EPC look-up so that EPC page is not shown and EPC-related details are not shown on the summary page, with the exception of address, as this leads to a bug where the address is blank on the summary page

### CODE REVIEW

File: help_to_heat/frontdoor/interface.py [hunk 1]
```diff
@@ -50,7 +51,7 @@ class FindAddressesSchema(marshmallow.Schema):
 
 
 class GetAddressSchema(marshmallow.Schema):
-    uprn = marshmallow.fields.Integer()
+    uprn = marshmallow.fields.String()
```
Comment (chidin194) [inline help_to_heat/frontdoor/interface.py:54]: This can actually go back to being an integer, I'll add a new class called `GetEpcSchema` and use accordingly

File: help_to_heat/frontdoor/interface.py [hunk 2]
```diff
@@ -452,11 +454,11 @@ def parse_dpa_to_address(self, dpa):
...
 class EPC(Entity):
     @with_schema(load=GetEPCSchema, dump=EPCSchema)
-    def get_epc(self, uprn, country):
+    def get_epc_scotland(self, uprn, country):
         try:
-            if country == "England" or country == "Wales":
-                epc = portal.models.EpcRating.objects.get(uprn=uprn)
-            elif country == "Scotland":
+            # if country == "England" or country == "Wales":
+            #     epc = portal.models.EpcRating.objects.get(uprn=uprn)
```
Comment (chidin194) [inline help_to_heat/frontdoor/interface.py:460]: This is only commented out for now as I wanted to leave it in as a reference to the `EpcRating` object which I'll need to remove everywhere

File: help_to_heat/frontdoor/interface.py [hunk 3]
```diff
@@ -467,8 +469,19 @@ def get_epc(self, uprn, country):
...
+        address_and_epc_details = data["data"]["assessments"]
+        return address_and_epc_details
+    
+    def get_epc_details(epc_rrn):
+        token = EPCApi.get_access_token()
+        data = EPCApi.get_epc_details(token, epc_rrn)
+        return data
 
-
+
```
Comment (chidin194) [inline help_to_heat/frontdoor/interface.py:484]: Will also address random line additions which aren't needed

File: help_to_heat/frontdoor/schemas.py [hunk 4]
```diff
@@ -702,7 +705,9 @@ class SessionSchema(Schema):
     postcode = fields.String(
         validate=validate.Regexp(postcode_regex_collection, error=_("Please enter a valid UK postcode"))
     )
-    uprn = fields.Integer()
+    uprn = fields.String()
+    rrn = fields.String()
+    epc_details = fields.Dict()
```
Comment (chidin194) [inline help_to_heat/frontdoor/schemas.py:710]: I think I still have some fields to add here, but changes will probably be ongoing until the work is complete. If my understanding is correct and this is just supposed to be a record of all the properties in the session data.

File: help_to_heat/frontdoor/views.py [hunk 5]
```diff
@@ -358,8 +360,50 @@ def save_data(self, request, session_id, page_name, *args, **kwargs):
 @register_page("address")
 class AddressView(PageView):
     def handle_post(self, request, session_id, page_name, data, is_change_page):
-        return redirect("frontdoor:page", session_id=session_id, page_name="address-select")
+        data = interface.api.session.get_answer(session_id, "address")
+        building_name_or_number = data["building_name_or_number"]
+        postcode = data["postcode"]
+        try:
+            interface.EPC.get_address_and_epc_rrn(building_name_or_number, postcode)
```
Comment (chidin194) [inline help_to_heat/frontdoor/views.py:367]: Could add a condition here to only make this call if the country isn't Scotland to decrease the number of API calls we make

File: help_to_heat/frontdoor/interface.py (hunk 1 above)
Comment (CheshireSwift) [reply help_to_heat/frontdoor/interface.py:54]: UPRNs should probably be strings anyway (same reasoning as something like phone numbers; it's made of digits, but we're not intending to do maths on it).

File: help_to_heat/frontdoor/views.py (hunk 5 above)
Comment (CheshireSwift) [reply help_to_heat/frontdoor/views.py:367]: Probably worthwhile, yeah.

File: help_to_heat/frontdoor/views.py [hunk 6]
```diff
@@ -396,9 +403,21 @@ def get_context(self, request, session_id, *args, **kwargs):
     def save_data(self, request, session_id, page_name, *args, **kwargs):
         rrn = request.POST["rrn"]
         epc = interface.EPC.get_epc_details(rrn)
+        print(epc)
```
Comment (CheshireSwift) [inline help_to_heat/frontdoor/views.py:406]: Probably don't want to keep this long-term 🙂

**Review by CheshireSwift — COMMENTED** [review]
What's currently here lgtm.

File: tests/test_eligibility.py [hunk 7]
```diff
@@ -357,19 +353,21 @@ def _do_test(country, council_tax_band, epc_rating):
     page = form.submit().follow()
 
     data = interface.api.session.get_answer(session_id, page_name="address")
-    assert data["building_name_or_number"] == "10"
-    assert data["postcode"] == "SW1A 2AA"
+    # assert data["building_name_or_number"] == "10"
+    # assert data["postcode"] == "SW1A 2AA"
```
Comment (jamiehumphries) [inline tests/test_eligibility.py:357]: Be careful not to commit commented out code! I'm assuming that these assertions need to be reinstated.

File: tests/test_eligibility.py [hunk 8]
```diff
@@ -357,19 +353,21 @@ def _do_test(country, council_tax_band, epc_rating):
...
+    form["rrn"] = "1111-1111-1111-1111-1111"
+
     page = form.submit().follow()
 
-    data = interface.api.session.get_answer(session_id, page_name="address-select")
-    assert data["uprn"] == 100023336956
-    assert data["address"] == "10, DOWNING STREET, LONDON, CITY OF WESTMINSTER, SW1A 2AA"
+    data = interface.api.session.get_answer(session_id, page_name="epc-select")
+    # assert data["rrn"] == "1111-1111-1111-1111-1111"
+    # assert data["address"] == "10, DOWNING STREET, LONDON, CITY OF WESTMINSTER, SW1A 2AA"
```
Comment (jamiehumphries) [inline tests/test_eligibility.py:366]: As above, please address these commented out lines.

**Review by jamiehumphries — CHANGES_REQUESTED** [review]
Most of the comments are mostly around style and code simplification.

However, there is a slightly bigger change to make, which is to replace the database storage of the access key with using the Django cache. We can discuss this one on Monday!

File: help_to_heat/frontdoor/interface.py [hunk 9]
```diff
@@ -468,6 +468,54 @@ def get_epc(self, uprn, country):
...
+    def get_epc_details(self, rrn):
+        return self.__try_with_authentication(lambda api: api.get_epc_details(rrn))
+
+    def __try_with_authentication(self, callback):
+        try:
+            token = self.__get_access_token()
+            return callback(EPCApi(token))
+        except requests.exceptions.RequestException as e:
+            status_code = e.response.status_code
+            if status_code == HTTPStatus.UNAUTHORIZED:
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/interface.py:485]: Yes! I believe the token expires after 3600 seconds, so an hour. I can refactor this logic, but might be worth discussing token refactoring more generally.

File: help_to_heat/frontdoor/mock_epc_api.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+__here__ = pathlib.Path(__file__).parent
+DATA_DIR = __here__ / "mock_epc_api_data"
+
+
+class MockEPCApi:
+    files = {
+        "address_and_rrn": "sample_address_and_rrn_response.json",
+        "epc": "sample_epc_response.json",
+        "access_token": "sample_token_response.json",
+    }
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/mock_epc_api.py:19]: This code is also present in `mock_os_api.py`, should I change it there too for consistency?

File: help_to_heat/frontdoor/epc_api.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+
+class EPCApi:
+    def __init__(self, token):
+        self.token = token
+
+    def get_access_token(self):
+        client_id = settings.EPC_API_CLIENT_ID
+        client_secret = settings.EPC_API_CLIENT_SECRET
+        token_url = "https://api.epb-staging.digital.communities.gov.uk/auth/oauth/token"
```
[the remark this answers is not in the document]
Comment (jamiehumphries) [reply help_to_heat/frontdoor/epc_api.py:16]: Thanks for checking. I hadn't seen `os_api.py`.

I think for the interest of getting this ticket sorted, let's just leave `os_api.py` alone.

I'll note it down as a potential area for tech debt cleanup, but there's probably plenty of competing things we'd want to do in the codebase that would be higher priority!

File: help_to_heat/frontdoor/interface.py [hunk 12]
```diff
@@ -67,7 +72,7 @@ class FullAddressSchema(marshmallow.Schema):
     address = marshmallow.fields.String()
 
 
-class GetEPCSchema(marshmallow.Schema):
+class GetScottishEPCSchema(marshmallow.Schema):
```
[the remark this answers is not in the document]
Comment (jamiehumphries) [reply help_to_heat/frontdoor/interface.py:75]: OK, that makes enough sense. Thanks for explaining. :)

File: help_to_heat/frontdoor/mock_epc_api.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+    def __init__(self, token):
+        self.token = token
+
+    def get_address_and_rrn(self, building, postcode):
+        content = (DATA_DIR / MockEPCApi.files["address_and_rrn"]).read_text()
+        return json.loads(content)
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/mock_epc_api.py:26]: I've added one in but just left it at the top-level, underneath where `DATA_DIR` is defined. I think this makes sense since it's a utility method, but I could be wrong!

[the remark this answers is not in the document]
Comment (jamiehumphries) [reply help_to_heat/frontdoor/mock_epc_api.py:26]: That sounds right to me. 👍

File: help_to_heat/frontdoor/views.py [hunk 14]
```diff
@@ -419,37 +537,70 @@ def get_context(self, request, session_id, *args, **kwargs):
...
+            uprn = session_data.get("uprn")
+            country = session_data.get("country")
+            uprn = session_data.get("uprn")
+            if uprn:
+                epc = interface.api.epc.get_epc_scotland(uprn, country)
+            else:
+                epc = {}
+            if not epc:
+                return redirect("frontdoor:page", session_id=session_id, page_name="benefits")
+            return super().handle_get(response, request, session_id, page_name, context)
```
[the remark this answers is not in the document]
Comment (jamiehumphries) [reply help_to_heat/frontdoor/views.py:590]: Re-reading this, if what I have above is correct, I think it can be further simplified to:
```
reference_number = session_data.get("uprn") if country == "Scotland" else session_data.get("rrn")
if reference_number:
    return super().handle_get(response, request, session_id, page_name, context)
else:
    return redirect("frontdoor:page", session_id=session_id, page_name="benefits")
```

File: help_to_heat/frontdoor/views.py [hunk 15]
```diff
@@ -419,37 +537,70 @@ def get_context(self, request, session_id, *args, **kwargs):
...
-        uprn = session_data.get("uprn")
         country = session_data.get("country")
-        uprn = session_data.get("uprn")
-        if uprn:
-            epc = interface.api.epc.get_epc(uprn, country)
+
+        if country == "Scotland":
+            uprn = session_data.get("uprn")
+            country = session_data.get("country")
+            uprn = session_data.get("uprn")
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/views.py:583]: Thanks for spotting this! Addressed.

File: help_to_heat/frontdoor/mock_epc_api.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/mock_epc_api.py:19]: Have used the filenames directly in the calls to the new load_test_response function :)

File: help_to_heat/frontdoor/interface.py [hunk 16]
```diff
@@ -451,15 +456,10 @@ def parse_dpa_to_address(self, dpa):
...
+    def get_epc_scotland(self, uprn):
         try:
-            if country == "England" or country == "Wales":
-                epc = portal.models.EpcRating.objects.get(uprn=uprn)
-            elif country == "Scotland":
-                epc = portal.models.ScottishEpcRating.objects.get(uprn=uprn)
-            else:
-                epc = None
+            epc = portal.models.ScottishEpcRating.objects.get(uprn=uprn)
         except (portal.models.EpcRating.DoesNotExist, portal.models.ScottishEpcRating.DoesNotExist):
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/interface.py:463]: Removed :)

File: help_to_heat/frontdoor/interface.py [hunk 17]
```diff
@@ -67,7 +72,7 @@ class FullAddressSchema(marshmallow.Schema):
     address = marshmallow.fields.String()
 
 
-class GetEPCSchema(marshmallow.Schema):
+class GetScottishEPCSchema(marshmallow.Schema):
     uprn = marshmallow.fields.Integer()
     country = marshmallow.fields.String()
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/interface.py:77]: Have removed `country` both from here and the `get_scottish_epc` function, along with its calls.

File: help_to_heat/frontdoor/views.py [hunk 18]
```diff
@@ -358,7 +360,107 @@ def save_data(self, request, session_id, page_name, *args, **kwargs):
...
+        return f"""{
+        a['address']['addressLine1'] + ', ' if a['address']['addressLine1'] else ''
+        }{
+        a['address']['addressLine2'] + ', ' if a['address']['addressLine2'] else ''
+        }{
+        a['address']['addressLine3'] + ', ' if a['address']['addressLine3'] else ''
+        }{
+        a['address']['addressLine4'] + ', ' if a['address']['addressLine4'] else ''
+        }{
+        a['address']['town']}, {a['address']['postcode']}"""
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/views.py:405]: Thanks, super helpful! That's much cleaner. This was a bit fiddly to implement!

File: help_to_heat/frontdoor/epc_api.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/epc_api.py:16]: @jamiehumphries 
Thanks! I'm adding in new variables to both the `web` and `web.template` files (with placeholder values), so this will be the full list, anticipating for what we might need in future:

```
EPC_API_BASE_URL_STAGING="https://api.epb-staging.digital.communities.gov.uk"
EPC_API_BASE_URL="https://api.epb.digital.communities.gov.uk"
EPC_API_CLIENT_ID_STAGING="client_id"
EPC_API_CLIENT_ID="client_id"
EPC_API_CLIENT_SECRET_STAGING="client_secret"
EPC_API_CLIENT_SECRET="client_secret"
```

File: help_to_heat/frontdoor/epc_api.py [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+
+        payload = {"grant_type": "client_credentials", "client_id": client_id, "client_secret": client_secret}
+
+        response = requests.post(token_url, data=payload)
+        response.raise_for_status()
+        return response.json().get("access_token")
+
+    def get_address_and_rrn(self, building, postcode):
+        url = f"https://api.epb-staging.digital.communities.gov.uk/api/assessments/domestic-epcs/search?postcode={postcode}&buildingNameOrNumber={building}"  # noqa E501
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/epc_api.py:25]: Implemented :)

File: help_to_heat/frontdoor/epc_api.py [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+        url = f"https://api.epb-staging.digital.communities.gov.uk/api/ecoplus/assessments/{rrn}"
+        return self.__api_call(url)
+
+    def __api_call(self, url):
+        headers = {"Authorization": f"Bearer {self.token}"}
+
+        response = requests.get(url, headers=headers)
+        response.raise_for_status()
+        data = response.json()
+        return data
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/epc_api.py:38]: Implemented :)

File: help_to_heat/frontdoor/views.py [hunk 21]
```diff
@@ -358,7 +360,107 @@ def save_data(self, request, session_id, page_name, *args, **kwargs):
...
+                    "epc_rating": "Not found",
+                    "accept_suggested_epc": "Not found",
+                    "epc_date": "",
+                },
+            )
+            return redirect("frontdoor:page", session_id=session_id, page_name="address-select")
+
+        address = self.format_address(epc["data"]["assessment"])
+
+        epc_details = epc["data"]["assessment"]
```
[the remark this answers is not in the document]
Comment (jamiehumphries) [reply help_to_heat/frontdoor/views.py:443]: You could still pass in `epc_details["address"]`, or even `epc_details.get("address") for more consistency with below. Not a major issue either way.

File: help_to_heat/frontdoor/views.py [hunk 22]
```diff
@@ -358,7 +360,107 @@ def save_data(self, request, session_id, page_name, *args, **kwargs):
...
+
+        if epc_details.get("uprn") is not None:
+            uprn = epc_details.get("uprn")
+        else:
+            uprn = ""
+
+        if epc_details.get("mainHeatingDescription") is not None:
+            property_main_heat_source = epc_details.get("mainHeatingDescription")
+        else:
+            property_main_heat_source = ""
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/views.py:453]: Refactored 😄

File: help_to_heat/frontdoor/views.py [hunk 23]
```diff
@@ -419,37 +537,70 @@ def get_context(self, request, session_id, *args, **kwargs):
...
+            address = session_data.get("address")
+            context = {}
+            if rrn:
+                epc = session_data.get("epc_details")
+            else:
+                epc = {}
+
+            context = {
+                "epc_rating": epc.get("currentEnergyEfficiencyBand").upper()
+                if epc.get("currentEnergyEfficiencyBand")
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/views.py:568]: Changed, thanks 😄

File: tests/test_eligibility.py (hunk 8 above)
Comment (chidin194) [reply tests/test_eligibility.py:366]: Apologies! These two assertions have been added back in.

File: tests/test_eligibility.py (hunk 7 above)
Comment (chidin194) [reply tests/test_eligibility.py:357]: I've removed these assertions as they no longer appear in the data associated with the answer. I'm not sure that's actually best practice. In `views.py`, in the handle_post method for the `AddressView`, the `building_name_or_number` and `postcode` that the user has provided are used to make a call to the EPC API; however, these two variables are overwritten by the response from the API:
```
        building_name_or_number = data["building_name_or_number"]
        postcode = data["postcode"]
        country = interface.api.session.get_answer(session_id, "country")["country"]
        try:
            if country == "Scotland":
                return redirect("frontdoor:page", session_id=session_id, page_name="address-select")
            else:
                address_and_rrn_details = interface.api.epc.get_address_and_epc_rrn(building_name_or_number, postcode)
                interface.api.session.save_answer(
                    session_id, page_name, {"address_and_rrn_details": address_and_rrn_details}
                )
```

Would it be better to keep all the data in there as a record of what the user has actually inputted on this page?

File: help_to_heat/frontdoor/views.py [hunk 24]
```diff
@@ -419,37 +537,70 @@ def get_context(self, request, session_id, *args, **kwargs):
...
-            epc = interface.api.epc.get_epc(uprn, country)
+
+        if country == "Scotland":
+            uprn = session_data.get("uprn")
+            address = session_data.get("address")
+            country = session_data.get("country")
+            if uprn:
+                epc = interface.api.epc.get_epc_scotland(uprn, country)
+            else:
+                epc = {}
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/views.py:549]: Changed to a ternary 😄

File: help_to_heat/frontdoor/views.py [hunk 25]
```diff
@@ -419,37 +537,70 @@ def get_context(self, request, session_id, *args, **kwargs):
...
-            "address": address,
-        }
-        return context
+            rrn = session_data.get("rrn")
+            address = session_data.get("address")
+            context = {}
+            if rrn:
+                epc = session_data.get("epc_details")
+            else:
+                epc = {}
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/views.py:564]: Changed to a ternary 😄

File: help_to_heat/frontdoor/views.py [hunk 26]
```diff
@@ -358,7 +360,107 @@ def save_data(self, request, session_id, page_name, *args, **kwargs):
...
+                return redirect("frontdoor:page", session_id=session_id, page_name="address-select")
+            else:
+                address_and_rrn_details = interface.api.epc.get_address_and_epc_rrn(building_name_or_number, postcode)
+                interface.api.session.save_answer(
+                    session_id, page_name, {"address_and_rrn_details": address_and_rrn_details}
+                )
+                return redirect("frontdoor:page", session_id=session_id, page_name="epc-select")
+        except Exception as e:  # noqa: B902
+            logger.exception(f"An error occurred: {e}")
+            interface.api.session.save_answer(
```
[the remark this answers is not in the document]
Comment (chidin194) [reply help_to_heat/frontdoor/views.py:377]: @jamiehumphries we effectively reset the test data as, if someone inputs address details for which the EPC API does find results, and then goes back and changes their address and no results are found, or they input it manually, old EPC details for the first address will still be in the session. I've extracted a top-level method for dealing with the resetting, but I do wonder if we should also reset details in the Scotland branch of the if statement. It's a very unlikely edge case, but just in case a user changes from an address in England (where there was a successful EPC look-up) to one in Scotland, where we don't perform EPC API lookups.

File: tests/test_eligibility.py (hunk 7 above)
Comment (jamiehumphries) [reply tests/test_eligibility.py:357]: Aha, thanks for explaining!

Looking into the tests further, the other relevant point is that you are adding back the other assertions (as discussed below):
```
    assert data["rrn"] == "1111-1111-1111-1111-1111"
    assert data["address"] == "10, DOWNING STREET, LONDON, CITY OF WESTMINSTER, SW1A 2AA"
```

Presumably the `address`, those would only be correctly populated if both the `building_name_or_number` and the `postcode` were correct. So we are testing the handling of those inputs anyway.

That being the case, I think removing these assertions is the correct thing to do, as you suggest. We don't need to keep the data around for any purpose.

File: help_to_heat/frontdoor/views.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (jamiehumphries) [reply help_to_heat/frontdoor/views.py:377]: Ah, I see. Thanks!

What you say makes sense. In fact, perhaps we should just always reset the data at the top of this `handle_post`? Regardless of any other factors, we are always going to want to discard the old data when handling a new post (I think).

[the remark this answers is not in the document]
Comment (CheshireSwift) [reply help_to_heat/frontdoor/views.py:377]: (I was going to react with a + but GitHub doesn't allow that.)

I agree that always resetting at the top of `handle_post` should be fine (with the benefit of slightly more familiarity with this code) 🙂

File: help_to_heat/frontdoor/epc_api.py [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+
+        payload = {"grant_type": "client_credentials", "client_id": client_id, "client_secret": client_secret}
+
+        response = requests.post(token_url, data=payload)
+        response.raise_for_status()
+        return response.json().get("access_token")
+
+    def get_address_and_rrn(self, building, postcode):
+        base_url = settings.EPC_API_BASE_URL
+        params = params = urllib.parse.urlencode({"postcode": postcode, "buildingNameOrNumber": building})
```
Comment (jamiehumphries) [inline help_to_heat/frontdoor/epc_api.py:28]: Typo:
```
params = params =
```
should just be:
```
params =
```

**Review by jamiehumphries — COMMENTED** [review]
Great stuff! Thanks for working through all this. 🙂
There's just one little typo that I spotted, but otherwise this is good to go.
